diff --git a/src/alphafold/common/confidence.py b/src/alphafold/common/confidence.py
index 68dca87..b2b96ec 100644
--- a/src/alphafold/common/confidence.py
+++ b/src/alphafold/common/confidence.py
@@ -23,6 +23,7 @@
 from typing import Dict, Optional, Tuple, Union, List
 import numpy as np
 import scipy.special
+import scipy.spatial
 import networkx as nx
 
 
diff --git a/src/alphafold/relax/amber_minimize.py b/src/alphafold/relax/amber_minimize.py
index 92d949d..000ace4 100644
--- a/src/alphafold/relax/amber_minimize.py
+++ b/src/alphafold/relax/amber_minimize.py
@@ -27,10 +27,10 @@ from alphafold.relax import utils
 import ml_collections
 import numpy as np
 import jax
-import openmm
-from openmm import unit
-from openmm import app as openmm_app
-from openmm.app.internal.pdbstructure import PdbStructure
+from simtk import openmm
+from simtk.openmm import unit
+from simtk.openmm import app as openmm_app
+from simtk.openmm.app.internal.pdbstructure import PdbStructure
 
 
 ENERGY = unit.kilocalories_per_mole
diff --git a/src/alphafold/relax/cleanup.py b/src/alphafold/relax/cleanup.py
index 3dd3160..32a1ce7 100644
--- a/src/alphafold/relax/cleanup.py
+++ b/src/alphafold/relax/cleanup.py
@@ -20,8 +20,8 @@ cases like removing chains of length one (see clean_structure).
 import io
 
 import pdbfixer
-from openmm import app
-from openmm.app import element
+from simtk.openmm import app
+from simtk.openmm.app import element
 
 def fix_pdb(pdbfile, alterations_info):
   """Apply pdbfixer to the contents of a PDB file; return a PDB string result.
